Core/src/Core/Filters/Algorithms/CalculateArrayHistogram.cpp

```diff
@@ -0,0 +1,315 @@
+      for(size_t i = 0; i < numOfElements; i++) // min and max in the input array
+      {
+        if(static_cast<float64>(m_InputArray[i]) > max)
+        {
+          max = static_cast<float64>(m_InputArray[i]);
+        }
+        if(static_cast<float64>(m_InputArray[i]) < min)
+        {
+          min = static_cast<float64>(m_InputArray[i]);
+        }
+      }
+      max = static_cast<float64>(max + 1); // ensure upper limit encapsulates max value
+      min = static_cast<float64>(min - 1); // ensure lower limit encapsulates min value
+    }
+    else
+    {
+      min = m_Min;
+      max = m_Max;
+    }
+
+    auto beginning = std::chrono::steady_clock::now();
+    const float64 increment = (max - min) / m_NumBins;
+    if(m_NumBins == 1) // if one bin, just set the first element to total number of points
+    {
+      m_Histogram[0] = max;
+      m_Histogram[1] = end;
+    }
+    else
+    {
+      for(usize i = start; i < end; i++)
+      {
+        m_Filter.updateThreadSafeProgress();
+        if(m_Filter.getCancel())
+        {
+          return;
+        }
+        const auto bin = std::floor((m_InputArray.at(i) - min) / increment);
+        if((bin >= 0) && (bin < m_NumBins))
+        {
+          m_Histogram[bin * 2 + 1]++;
+        }
+        else
+        {
+          m_Overflow++;
+        }
+      }
+    }
+
+    for(int64 i = 0; i < m_NumBins; i++)
+    {
+      m_Histogram[(i * 2)] = static_cast<float64>(min + (increment * (i + 1))); // load bin maximum into respective postion {(x, ), (x , ), ...}
+    }
+  }
+
+  void operator()(const Range& range) const
+  {
+    generate(range.min(), range.max());
+  }
+
+private:
+  CalculateArrayHistogram& m_Filter;
+  const int32 m_NumBins = 1;
+  float64 m_Min = 0.0;
+  float64 m_Max = 0.0;
+  const DataArray<T>& m_InputArray;
+  Float64Array& m_Histogram;
+  std::atomic<usize>& m_Overflow;
+};
+} // namespace
+
+// -----------------------------------------------------------------------------
+CalculateArrayHistogram::CalculateArrayHistogram(DataStructure& dataStructure, const IFilter::MessageHandler& mesgHandler, const std::atomic_bool& shouldCancel,
+                                                 CalculateArrayHistogramInputValues* inputValues)
+: m_DataStructure(dataStructure)
+, m_InputValues(inputValues)
+, m_ShouldCancel(shouldCancel)
+, m_MessageHandler(mesgHandler)
+{
+}
+
+// -----------------------------------------------------------------------------
+CalculateArrayHistogram::~CalculateArrayHistogram() noexcept = default;
+
+// -----------------------------------------------------------------------------
+void CalculateArrayHistogram::updateProgress(const std::string& progMessage)
+{
+  m_MessageHandler({IFilter::Message::Type::Info, progMessage});
+}
+// -----------------------------------------------------------------------------
+void CalculateArrayHistogram::updateThreadSafeProgress()
```

Comment:
See comments above about using ApplyTransformationToGeometry as a design pattern for this class.